sdk/storage/azure-storage-data-movement/src/main/java/com/azure/storage/data/movement/PathScanner.java

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.storage.data.movement;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import com.azure.core.util.logging.ClientLogger;
+import reactor.core.scheduler.Schedulers;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.stream.BaseStream;
+
+/**
+ * Scanner for enumerating local files/folders at a specified location through publishers.
+ *
+ * TODO: Replace placeholder Javadoc
+ */
+public final class PathScanner {
+    private final ClientLogger logger = new ClientLogger(PathScanner.class);
+    private final String basePath;
+
+    /**
+     * Constructor for {@link PathScanner}.
+     *
+     * TODO: Replace placeholder Javadoc
+     *
+     * @param path The local path to be scanned, either relative to execution location or absolute.
+     * @throws IllegalArgumentException if the provided path does not point to anything.
+     */
+    public PathScanner(String path) {
+        try {
+            // Resolve the given path to an absolute path in case it isn't one already
+            Path pathObj = Paths.get(path);
+            basePath = pathObj.toAbsolutePath().toString();
+
+            // Check if the path exists; throw an error if there's nothing
+            // present at the given path
+            if (!(Files.isDirectory(pathObj) || Files.isRegularFile(pathObj))) {
+                throw new IllegalArgumentException(String.format("No item(s) located at the path '%s'.", basePath));
+            }
+        } catch (RuntimeException e) {
+            // Possible throws: nonexistent item, malformed path
+            throw logger.logExceptionAsError(e);
+        }
+    }
+
+    /**
+     * Enumerates the files/folders at the path this scanner points to using a non-blocking
+     * publisher.
+     *
+     * TODO: Replace placeholder Javadoc comment
+     *
+     * @param skipSubdirectories Sets whether to continue enumerating with warnings or throw through the
+     *                           Flux when a folder can't be accessed.
```

Comment:
What does skipSubdirectories do ? Are we attempting to recursively scan if this is false ?

Reply:
If it's true, then if any subfolders during enumerations can't be opened due to read errors or something else, the scanner will log it as a warning and continue trying to enumerate the rest of its queue; otherwise, it logs an error and passes it into the Flux in such a case, ending enumeration.

Reply:
I can see that from the code. Question is in what scenario I'd pass true vs false as a caller of this method?

Reply:
I think a customer could want to pass true and do this if they don't want the whole job to stop once the first error is encountered. Is that a correct understanding @rnpmsft? skipSubDirectories might not be the best name for this if that is the intent here.

Reply:
If I'm aware that a folder contains some contents that are inaccessible but not critical to the current upload, and would like to send up everything else, then I could pass in true.

Reply:
@gapra-msft Yup, pretty much. Any suggestions as to what to name it? Maybe `continueOnError`?

Reply:
yes, continueOnError is better.

Reply:
Yes, that is an improvement from the current name, and with good documentation can convey the intent well. @kasobol-msft thoughts on the new name?